compiler-builtins/src/thumbv6k.rs

```diff
@@ -0,0 +1,326 @@
+                // casts are okay because $ty and $base implement the same layout.
+                unsafe { <$base as Atomic>::load_relaxed(src.cast::<$base>()) as Self }
+            }
+            #[inline]
+            unsafe fn cmpxchg(dst: *mut Self, old: Self, new: Self) -> Self {
+                // SAFETY: the caller must uphold the safety contract.
+                // casts are okay because $ty and $base implement the same layout.
+                unsafe {
+                    <$base as Atomic>::cmpxchg(dst.cast::<$base>(), old as $base, new as $base)
+                        as Self
+                }
+            }
+        }
+    };
+}
+delegate_signed!(i8, u8);
+delegate_signed!(i16, u16);
+delegate_signed!(i32, u32);
+
+// Generic atomic read-modify-write operation
+//
+// We could implement RMW more efficiently as an assembly LL/SC loop per operation,
+// but we won't do that for now because it would make the implementation more complex.
+//
+// We also do not implement LL and SC as separate functions. This is because it
+// is theoretically possible for the compiler to insert operations that might
+// clear the reservation between LL and SC. See https://github.com/taiki-e/portable-atomic/blob/58ef7f27c9e20da4cc1ef0abf8b8ce9ac5219ec3/src/imp/atomic128/aarch64.rs#L44-L55
+// for more details.
+unsafe fn atomic_rmw<T: Atomic, F: Fn(T) -> T, G: Fn(T, T) -> T>(ptr: *mut T, f: F, g: G) -> T {
+    loop {
+        // FIXME(safety): preconditions review needed
+        let curval = unsafe { T::load_relaxed(ptr) };
+        let newval = f(curval);
+        // FIXME(safety): preconditions review needed
+        if unsafe { T::cmpxchg(ptr, curval, newval) } == curval {
+            return g(curval, newval);
+        }
+    }
+}
+
+macro_rules! atomic_rmw {
+    ($name:ident, $ty:ty, $op:expr, $fetch:expr) => {
+        intrinsics! {
+            pub unsafe extern "C" fn $name(ptr: *mut $ty, val: $ty) -> $ty {
+                // FIXME(safety): preconditions review needed
+                unsafe {
+                    atomic_rmw(
+                        ptr,
+                        |x| $op(x as $ty, val),
+                        |old, new| $fetch(old, new)
+                    ) as $ty
+                }
+            }
+        }
+    };
+
+    (@old $name:ident, $ty:ty, $op:expr) => {
+        atomic_rmw!($name, $ty, $op, |old, _| old);
+    };
+
+    (@new $name:ident, $ty:ty, $op:expr) => {
+        atomic_rmw!($name, $ty, $op, |_, new| new);
+    };
+}
+macro_rules! atomic_cmpxchg {
+    ($name:ident, $ty:ty) => {
+        intrinsics! {
+            pub unsafe extern "C" fn $name(ptr: *mut $ty, oldval: $ty, newval: $ty) -> $ty {
+                // FIXME(safety): preconditions review needed
+                unsafe { <$ty as Atomic>::cmpxchg(ptr, oldval, newval) }
+            }
+        }
+    };
+}
+
+atomic_rmw!(@old __sync_fetch_and_add_1, u8, |a: u8, b: u8| a.wrapping_add(b));
```

Comment:
The macro calls from here to L319 are the same as in `arm_linux.rs`  (the definitions of the called macros are different), so I think it would be preferable to extract them into a separate file.